api/service/syncing/syncing.go

```diff
@@ -770,28 +784,60 @@ Loop:
 			currentHeight := bc.CurrentBlock().NumberU64()
 
 			if currentHeight >= otherHeight {
-				utils.Logger().Info().
-					Msgf("[SYNC] Node is now IN SYNC! (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
-						isBeacon, bc.ShardID(), otherHeight, currentHeight)
+				utils.Logger().Debug().
+					Bool("isBeacon", isBeacon).
+					Uint32("ShardID", bc.ShardID()).
+					Uint64("otherHeight", otherHeight).
+					Uint64("currentHeight", currentHeight).
+					Msg("[SYNC] Node is now IN SYNC!")
 				break Loop
 			} else {
 				utils.Logger().Debug().
-					Msgf("[SYNC] Node is Not in Sync (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
-						isBeacon, bc.ShardID(), otherHeight, currentHeight)
+					Bool("isBeacon", isBeacon).
+					Uint32("ShardID", bc.ShardID()).
+					Uint64("otherHeight", otherHeight).
+					Uint64("currentHeight", currentHeight).
+					Msg("[SYNC] Node is Not in Sync.")
 			}
 			startHash := bc.CurrentBlock().Hash()
 			size := uint32(otherHeight - currentHeight)
 			if size > SyncLoopBatchSize {
 				size = SyncLoopBatchSize
 			}
-			err := ss.ProcessStateSync(startHash[:], size, bc, worker)
-			if err != nil {
-				utils.Logger().Error().Err(err).
-					Msgf("[SYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
-						isBeacon, bc.ShardID(), otherHeight, currentHeight)
-				// should we still call UpdateConsensusInformation() upon state sync failure?
-				// how to handle error here?
+
+			retryCount := 0
+			var err error
+			for {
+				err = ss.ProcessStateSync(startHash[:], size, bc, worker)
+				if err == nil {
+					break
+				}
+				if retryCount >= stateSyncRetryLimit {
+					utils.Logger().Warn().Err(err).
+						Bool("isBeacon", isBeacon).
+						Uint32("ShardID", bc.ShardID()).
+						Uint64("otherHeight", otherHeight).
+						Uint64("currentHeight", currentHeight).
+						Msg("[SYNC] ProcessStateSync failed after retries. Node failed to sync.")
+					break
+				}
+
+				utils.Logger().Warn().Err(err).
+					Bool("isBeacon", isBeacon).
+					Uint32("ShardID", bc.ShardID()).
+					Uint64("otherHeight", otherHeight).
+					Uint64("currentHeight", currentHeight).
+					Int("retryCount", retryCount).
+					Msg("[SYNC] ProcessStateSync failed. Retrying ...")
+
+				ss.syncConfig.ForEachPeer(func(configPeer *SyncPeerConfig) (brk bool) {
```

Comment:
make this anonymous function outside of loop, otherwise we keep making it freshly over and over. and what is purpose of brk? the function always gives back true so no purpose of have a return value at all it seems

Reply:
@fxfactorial described in the comment of this function here

`// ForEachPeer calls the given function with each peer.
// It breaks the iteration iff the function returns true.
func (sc *SyncConfig) ForEachPeer(f func(peer *SyncPeerConfig) (brk bool)) {
	sc.mtx.RLock()
	defer sc.mtx.RUnlock()
	for _, peer := range sc.peers {
		if f(peer) {
			break
		}
	}
}`

Reply:
Gotcha , let’s still make the cb just once , no need to stack thrash